# Tarea 1 Bases de Datos 2023-1 Grupo XX



## Integrantes:
* Gustav Jacobsson - 202290241-8
* Nombre Integrante 2 - Rol-2

## Requisitos para Correr el Notebook

Para poder utilizar este Notebook se debe tener instalado:
- **Python 3**
- **Jupyter**
- La librería **ipython-sql** (Poder ejecutar sql en Jupyter Notebook)
- PostgreSQL y su conector desde Python(**psycopg2**)

**IMPORTANTE:** Este Jupyter Notebook se conectará a su servicio de Base de Datos PostgreSQL de manera **local**, por lo que **no funcionará correctamente en el entorno Google Collab.**

Para instalar las dependencias, ejecute (**recuerde actualizar pip3**):

In [4]:
! pip3 install ipython-sql
! pip3 install psycopg2

#Use to generate names
! pip3 install names


[notice] A new release of pip available: 22.2.2 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip available: 22.2.2 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip available: 22.2.2 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


## Importar dependencias

In [5]:
from datetime import datetime
import random as r
import names

#Librerías para PostgreSQL
import psycopg2

Para conectarse a la base de datos, completen los datos relevantes en la variable DATABASE_URL.

* Cambie `user` por el usuario que utilizó en Postgres para acceder a la Base de Datos. Por defecto `postgres`.
* Cambie `password` por la contraseña de dicho usuario. **Recuerde borrarla antes de entregar su tarea.**
* Cambie `port` por el puerto en que se ejecuta su servicio de Base de Datos. Por defecto `5432`.
* Cambie `database` por el nombre de la BD (Schema) que creó en Postgres.

Ejemplo: `postgresql://postgres:pass_segura@localhost:5432/tarea-bd`

In [6]:
%reload_ext sql
%env DATABASE_URL=postgresql://postgres:papelucho1@localhost:5432/Tarea1DB

env: DATABASE_URL=postgresql://postgres:papelucho1@localhost:5432/Tarea1DB


Para conectarse de forma declarativa a la Base de Datos, utilice el siguiente código:

In [7]:
#Recuerde cambiar los valores de USER, DBNAME y PASSWORD por los mismos de la celda superior.
connection = psycopg2.connect(
   host="localhost",
   user="postgres",
   dbname='tarea1',
   password="Jageascool12!"
)

print(connection)
cursor = connection.cursor()

OperationalError: 

## Crear tablas (36 pts.)

Aquí inserta el código SQL que creara las tablas de la base de datos:

### Tabla "facciones" (6 pts)

In [13]:
%%sql CREATE TABLE facciones(
	id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
	nombre VARCHAR(45) NOT NULL,
	descripcion VARCHAR(45) NOT NULL
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "trabajos" (6 pts)

In [14]:
%%sql CREATE TABLE trabajos(
	id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
	nombre VARCHAR(45) NOT NULL,
	descripcion VARCHAR(45) NOT NULL,
    sueldo INTEGER NOT NULL	
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "personas" (6 pts)

In [16]:
%%sql CREATE TABLE personas(
	id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
	nombre VARCHAR(45) NOT NULL,
    apellidos VARCHAR(45) NOT NULL,
    fecha_nacimiento DATE NOT NULL,
	infectado BOOL NOT NULL,
    conyugue INTEGER,
    FOREIGN KEY (conyugue) REFERENCES personas (id)
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "cuentas_bancarias" (6 pts)

In [17]:
%%sql CREATE TABLE cuentas_bancarias(
	id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
	tipo_de_cuenta VARCHAR(45) NOT NULL,
	banco VARCHAR(45) NOT NULL,
    saldo INTEGER NOT NULL,
    id_persona INTEGER NOT NULL,
    FOREIGN KEY (id_persona) REFERENCES personas (id)
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "persona_pertenece_faccion" (6 pts)

In [18]:
%%sql CREATE TABLE persona_pertenece_faccion(
	id_persona INTEGER,
    id_faccion INTEGER,
    FOREIGN KEY (id_persona) REFERENCES personas(id),
    FOREIGN KEY (id_faccion) REFERENCES facciones(id),
	PRIMARY KEY(id_persona,id_faccion)
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "persona_tiene_trabajo" (6 pts)

In [25]:
%%sql CREATE TABLE persona_tiene_trabajo(
	id_persona INTEGER,
    id_trabajo INTEGER,
    estado BOOL NOT NULL,
    ultima_vez_realizado TIMESTAMP NOT NULL,
    FOREIGN KEY (id_persona) REFERENCES personas(id)
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

## Código Generador de Datos

Ejecute este código para cargar datos a las tablas creadas.

PD: Si el codigo tira error por "violar la unicidad de las PK" es por una coincidencia de numeros random, solamente volver a ejecutar el codigo

In [26]:
#Recuerde haber ejecutado todas las celdas anteriores!

# Generador de timestamp aleatorio.
def generate_timestamp_nacimiento():
    year = r.randint(1970,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

def generate_timestamp_trabajo():
    year = r.randint(2019,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

print("Borrando Datos Antiguos...")

# Limpia la base de datos.
connection.commit() # En caso de algun error
query = 'TRUNCATE TABLE "facciones", "trabajos", "personas", "cuentas_bancarias", "persona_pertenece_faccion", "persona_tiene_trabajo" RESTART IDENTITY'
cursor.execute(query)
connection.commit()

print("Generando Datos...")

facciones = [("FEDRA", "La Agencia Federal de Respuesta a Desastres"), ("Luciérnagas", "Grupo de milicias revolucionario"), ("Carroñeros", "Sin identificación"), ("Pythonia", "Grupo defensor de la libertad"), ("MaCdonia", "Los exiliados de la carceles"), ("JavaScriptia", "Grupo defensor de la paz")]
trabajos = [("Limpieza", "Encargado de limpiar las calles"),
("Limpiador de cloacas","Encargado de limpiar las cloacas"),
("Militar","Encargado de proteger la ciudad"),
("Explorador","Encargado de explorar la ciudad"),
("Médico","encargado de curar a los enfermos"),
("Profesor","Encargado de enseñar a los niños"),
("Cremador","Encargado de quemar los cadáveres"),
("Científico","Encargado de investigar"),
("Cocinero","Encargado de cocinar"),
("Carpintero","Encargado de construir"),
("Cerrajero","Encargado de reparar las cerraduras"),
("Cartero","Encargado de repartir los correos"),
("Cajero","Encargado de cobrar los impuestos")]
tipos_de_cuenta_bancaria = ["Cuenta Vista", "Cuenta Ahorro", "Cuenta Corriente", "Cuenta RUT", "Cuenta de Inversion", "Cuenta PrePago", "Cuenta de Credito"]
bancos = ["FedraBank S.A", "Banco Pedrito Pascal SpA", "Banco Pythonia S.A", "Banco de la Nación S.A", "Banco Santander SpA", "Banco BCI SpA", "Banco Estado SpA", "Banco Consorcio S.A", "Banco BBVA", "Banco Scotiabank"]

# Variables.
n_facciones = len(facciones)
n_trabajos = len(trabajos)
n_personas = 1000
n_cuentas_bancarias = 120
n_persona_pertenece_faccion = 300
n_persona_tiene_trabajo = 300
n_casados = 100

# Genera personas.
personas = []
for _ in range(n_personas):
    nombre = names.get_first_name()
    apellidos = names.get_last_name()+" "+names.get_last_name()
    fecha_nacimiento = generate_timestamp_nacimiento()
    infectado = r.choice([True, False])
    
    personas.append((nombre, apellidos, fecha_nacimiento, infectado, None))


# Genera cuentas bancarias.
cuentas_bancarias = []
for _ in range(n_cuentas_bancarias):
    tipo_de_cuenta = r.choice(tipos_de_cuenta_bancaria)
    banco = r.choice(bancos)
    saldo = r.randint(0, 10000)
    id_persona = r.choice(range(1, n_personas+1))
    
    cuentas_bancarias.append((tipo_de_cuenta, banco, saldo, id_persona))
    
    
# Generate persona_pertenece_faccion.
persona_pertenece_faccion = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_facciones+1))) for _ in range(1, n_persona_pertenece_faccion+1)])

# Generate persona_tiene_trabajo.
persona_tiene_trabajo_prev = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_trabajos+1))) for _ in range(1, n_persona_tiene_trabajo+1)])
persona_tiene_trabajo = [(id_persona, id_trabajo, r.choice([True, False]), generate_timestamp_trabajo()) for id_persona, id_trabajo in persona_tiene_trabajo_prev]
    
# SQL Statements.
inserts = [
    'INSERT INTO "facciones" ("nombre", "descripcion") VALUES (%s, %s);',
    'INSERT INTO "trabajos" ("nombre", "descripcion", "sueldo") VALUES (%s, %s, %s);',
    'INSERT INTO "personas" ("nombre", "apellidos", "fecha_nacimiento", "infectado", "conyugue") VALUES (%s, %s, %s, %s, %s);',
    'INSERT INTO "cuentas_bancarias" ("tipo_de_cuenta", "banco", "saldo", "id_persona") VALUES (%s, %s, %s, %s);',
    'INSERT INTO "persona_pertenece_faccion" ("id_persona", "id_faccion") VALUES (%s, %s);',
    'INSERT INTO "persona_tiene_trabajo" ("id_persona", "id_trabajo", "estado", "ultima_vez_realizado") VALUES (%s, %s, %s, %s);'
]

for faccion in facciones:
    cursor.execute(inserts[0], faccion)

for trabajo in trabajos:
    t = (trabajo[0], trabajo[1], r.randint(100, 5000))
    cursor.execute(inserts[1], t)

# Crear una lista con personas casadas y otra con personas solteras. donde se cambie la ide de la persona que se casó.

for persona in personas:
    cursor.execute(inserts[2], persona)
    
for cb in cuentas_bancarias:
    cursor.execute(inserts[3], cb)

for ppf in persona_pertenece_faccion:
    cursor.execute(inserts[4], ppf)
    
for ptt in persona_tiene_trabajo:
    cursor.execute(inserts[5], ptt)
    
connection.commit()


# Generar Matrimonios
n_repetidos = []

for i in range(n_casados):
    random = r.randint(1, n_personas)
    if random in n_repetidos:
        continue
    else:
        random2 = r.randint(1, n_personas)
        if random2 in n_repetidos:
            continue
        else:
            n_repetidos.append(random)
            n_repetidos.append(random2)
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random2, random))
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random, random2))
            connection.commit()

print("Todo Listo!")

Borrando Datos Antiguos...
Generando Datos...
Todo Listo!


## Consultas (64 pts.)

Escriba aquí las consultas pedidas. Recuerde agregar una descripción _breve_ de cómo funciona la solución. **Escriba las consultas de forma ordenada.**

**NOTA:** Cuando en una consulta diga "X cosa" significa que ustedes deben de elegir ese X por lo que para cualquier otro X debe de funcionar de igual forma.

### Consulta 1 (6 pts)
**ID, nombre y primer apellido de los habitantes de FEDRA**

Ejemplo:

|ID |  Nombre      |  Apellido  |
|---|-----------|-----------|
| 12 |Joel | Miller |
|21 | Ellie | Williams |
| ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._


In [36]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    personas.id, personas.nombre, SPLIT_PART(personas.apellidos, ' ' ,1) AS apellidos
FROM 
    personas JOIN persona_pertenece_faccion ON personas.id = persona_pertenece_faccion.id_persona JOIN facciones ON persona_pertenece_faccion.id_faccion = facciones.id
WHERE
    persona_pertenece_faccion.id_faccion = 1;

 * postgresql://postgres:***@localhost:5432/Tarea1DB
57 rows affected.


[(21, 'Margarita', 'Thompson'),
 (33, 'Marian', 'Guardado'),
 (59, 'Brenda', 'Mehtani'),
 (32, 'Aurora', 'Lynch'),
 (132, 'John', 'Fitzgerald'),
 (137, 'Marie', 'Mcallen'),
 (172, 'Sarah', 'Rodriguez'),
 (220, 'Travis', 'Rock'),
 (222, 'Michele', 'Sitter'),
 (231, 'Melissa', 'Perry'),
 (185, 'Cynthia', 'Vanbuskirk'),
 (264, 'Salena', 'Ratterman'),
 (312, 'Peter', 'Guerra'),
 (329, 'Maria', 'Brobst'),
 (321, 'Lisa', 'Olson'),
 (379, 'Monty', 'Debois'),
 (394, 'Wm', 'Butts'),
 (406, 'William', 'Bowen'),
 (437, 'Peter', 'Butler'),
 (439, 'Brad', 'Snowden'),
 (485, 'Robert', 'Chapman'),
 (423, 'Joan', 'Harris'),
 (407, 'Michael', 'Perry'),
 (517, 'Colleen', 'Graves'),
 (524, 'Eric', 'Curtis'),
 (533, 'Frank', 'Olsen'),
 (544, 'Keith', 'Zimmerman'),
 (581, 'Latrice', 'Mcelroy'),
 (566, 'James', 'Reed'),
 (637, 'Michelle', 'Lombardo'),
 (645, 'Chelsea', 'Creekmore'),
 (652, 'Stacey', 'Moseley'),
 (663, 'Daphne', 'Jones'),
 (687, 'Kathryn', 'Gray'),
 (707, 'Donna', 'Morvillo'),
 (710, 'Ronda', 'Sartoris'),
 (715, 'Joel', 'Henley'),
 (721, 'Helena', 'Liberti'),
 (729, 'John', 'Wheaton'),
 (746, 'Johnny', 'White'),
 (759, 'Michael', 'Simmons'),
 (832, 'Paul', 'Feuerborn'),
 (864, 'Sandra', 'Patten'),
 (869, 'Eugene', 'Harold'),
 (872, 'Robert', 'Glass'),
 (824, 'Christine', 'Thomas'),
 (889, 'James', 'Gutierrez'),
 (890, 'Marjorie', 'Tillman'),
 (908, 'John', 'Kilmer'),
 (909, 'Wilbert', 'Weaver'),
 (918, 'Kathy', 'Jordan'),
 (931, 'Tonya', 'Hoskins'),
 (986, 'Marlon', 'Hale'),
 (887, 'Nicholas', 'Plaza'),
 (927, 'Herbert', 'Johnson'),
 (894, 'Christine', 'Chambers'),
 (424, 'Jeannette', 'Cornish')]

### Consulta 2 (6 pts)
**Cantidad de personas nacidas después del inicio de la pandemia.**(inicio de la pandemia 01-12-2019)

Ejemplo:


| Cantidad     | 
|--------------|
| 292 | 

**Solución:** _Escriba aquí la descripción de su solución._

In [37]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    COUNT(personas.fecha_nacimiento) AS Cantidad
FROM 
    personas
WHERE
    personas.fecha_nacimiento > '2019-12-01'

 * postgresql://postgres:***@localhost:5432/Tarea1DB
1 rows affected.


[(44,)]

### Consulta 3 (6 pts)
**ID, nombre, apellidos y saldo total de todas las personas**( Limitar a 20 filas para visualizar)

Ejemplo:

| ID    | Nombre | Apellido | Dinero | 
|----------------|--------|-|-|
| 13 | Tommy | Miller | 12501 |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [38]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    personas.id, personas.nombre, SPLIT_PART(personas.apellidos, ' ' ,1), cuentas_bancarias.saldo
FROM 
    personas JOIN cuentas_bancarias ON personas.id = id_persona
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/Tarea1DB
20 rows affected.


[(8, 'Vernon', 'Cliff', 2567),
 (11, 'Patrick', 'Hill', 4014),
 (28, 'Ashley', 'Hancock', 8733),
 (53, 'Melvin', 'Barbee', 6762),
 (56, 'Bruce', 'Reed', 8652),
 (62, 'Stephanie', 'Evans', 2662),
 (63, 'David', 'Griffin', 4646),
 (64, 'Lana', 'Kelly', 643),
 (64, 'Lana', 'Kelly', 4146),
 (80, 'Holly', 'Stewart', 8327),
 (88, 'Rebecca', 'Helfer', 2112),
 (93, 'Harry', 'Denmon', 6270),
 (94, 'Andrew', 'Grimes', 5405),
 (97, 'John', 'Bond', 7836),
 (104, 'Aaron', 'Ward', 9804),
 (105, 'Shannon', 'Schumacher', 5393),
 (108, 'Guadalupe', 'Hanford', 171),
 (111, 'John', 'Turpen', 1830),
 (114, 'Gerald', 'Dorn', 3891),
 (4, 'Lily', 'Bowen', 6779)]

### Consulta 4 (6 pts)
**Nombre de la Facción y la cantidad de trabajadores activos hasta fecha X.**

Ejemplo:

| Facción     | Cantidad de trabajadores |
|----------------|--------|
| FEDRA | 1273 |
| Luciérnagas | 121 |
| Carroñeros  | 72 |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [42]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    facciones.nombre, COUNT(persona_tiene_trabajo.estado)
FROM 
    persona_pertenece_faccion JOIN personas ON persona_pertenece_faccion.id_persona = personas.id JOIN persona_tiene_trabajo ON personas.id = persona_tiene_trabajo.id_persona
WHERE
    persona_tiene_trabajo.estado = true;


 * postgresql://postgres:***@localhost:5432/Tarea1DB
(psycopg2.errors.UndefinedTable) falta una entrada para la tabla «facciones» en la cláusula FROM
LINE 4:     facciones.nombre, COUNT(persona_tiene_trabajo.estado)
            ^

[SQL: /* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    facciones.nombre, COUNT(persona_tiene_trabajo.estado)
FROM 
    persona_pertenece_faccion JOIN personas ON persona_pertenece_faccion.id_persona = personas.id JOIN persona_tiene_trabajo ON personas.id = persona_tiene_trabajo.id_persona
WHERE
    persona_tiene_trabajo.estado = true;]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Consulta 5 (6 pts)
**Nombre de la Facción y su cantidad de personas infectadas.**

Ejemplo:

|  Facción  | Personas infectadas | 
|--------|-------|
| FEDRA | 251 |
| Luciérnagas | 17 |
| Carroñeros  | 12 |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    * 
FROM 
    *

### Consulta 6 (6 pts)
**Nombre de la Facción y su trabajo más realizado.**

Ejemplo:

|  Facción  | Trabajo | 
|--------|-------|
| FEDRA | Limpiador de cloacas |
| Luciérnagas | Médico |
| Carroñeros  | Explorador |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql


SELECT
   facciones.nombre, trabajos.nombre
FROM facciones
INNER JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion
INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id
INNER JOIN persona_tiene_trabajo ON personas.id = persona_tiene_trabajo.id_persona
INNER JOIN trabajos ON persona_tiene_trabajo.id_trabajo = trabajos.id
GROUP BY facciones.nombre
ORDER BY COUNT(*) DESC;



### Consulta 7 (8 pts)
**Nombre de la Facción y su Saldo total** (Es decir, la suma total de los saldos de cada persona perteneciente a la facción.)

Ejemplo:

|Facción | Saldo |
|--------|-|
| FEDRA | 523000 |
| Luciérnagas | 130023 |
| Carroñeros  | 17432 |
| ...  | ...|

**Solución:** Each faction is INNER JOINED with "perosna_pertenece_faccion" and then with "personas" and lastly with "cuenta_bancaria". These are then grouped by faction, and "SUM(cuenta_bancaria.sueldo) is displayed._

In [ ]:
%%sql


SELECT
    facciones.nombre, SUM(cuentas_bancarias.saldo) AS saldo_total
FROM facciones
    INNER JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion
    INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id
    INNER JOIN cuentas_bancarias ON personas.id = cuentas_bancarias.id_persona
    GROUP BY facciones.nombre;

### Consulta 8 (6 pts)
**Nombre del Banco, cantidad de usuarios pertenecientes al banco y saldo total del banco.**

###### Es decir, todos los bancos que ocupan los habitantes de cada faccion

Ejemplo:

|Banco | Cantidad|Saldo|
|----|---|---|
|FedraBank S.A |321|20000|
|Banco Pedrito Pascal SpA  |21|200| 
| Banco De Chile|213|56000|
| ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

SELECT
    cuentas_bancarias.banco, COUNT(*), SUM(cuentas_bancarias.saldo)
FROM cuentas_bancarias
GROUP BY cuentas_bancarias.banco

### Consulta 9 (8 pts)
**Nombre de la Facción y el Banco más utilizado por sus habitantes**(Si hay empate, que se muestre solo uno.)

Ejemplo:
   
|Facción | Banco |
| - | - |
| FEDRA | FedraBank S.A |
| Luciérnagas | Banco Pedrito Pascal SpA |
| ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

SELECT
facciones.nombre, cuentas_bancarias.banco as banco_mas_popular
FROM facciones
INNER JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion
INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id
INNER JOIN cuentas_bancarias ON personas.id = cuentas_bancarias.id_persona
GROUP BY facciones.nombre, banco_mas_popular
ORDER BY COUNT(*) DESC;

### Consulta 10 (6 pts)
**ID, nombre y segundo apellido de las personas traidoras**(Es decir, si la persona pertenece a dos o más facciones.)

Ejemplo:
   
|ID | Nombre | Apellido | 
|-|-------|-|
| 1 | Pedro | Pascal |
| 2 | Bella | Ramsey |
| ... | ... | ... | 

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

SELECT p.id as "ID", p.nombre, SPLIT_PART(p.apellidos, ' ', 1) AS apellido
FROM personas p
INNER JOIN persona_pertenece_faccion ON p.id = persona_pertenece_faccion.id_persona
GROUP BY p.id, p.nombre, SPLIT_PART(p.apellidos, ' ', 1)
HAVING COUNT(persona_pertenece_faccion.id_faccion) > 1;

### Consulta BONUS (6 pts.)

**Nombre y Apellidos de las personas casadas**(solo se muestra una única vez la pareja)

Ejemplo:
   
|Nombre persona 1 | Apellidos persona 1 | Nombre persona 2 | Apellidos persona 2 |
|-|-------|-|-|
| Felipe | Martinez | Raul | Cuello |
| Rodrigo | Munita | Juan | Mira |
| Nicole | Rojel | Benjamin | Campos |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *